### PR TITLE
[23189] - enable custom SSL cipher suites in jaxrs client requests

### LINKED ISSUE #22738: SSLContext defined in ClientBuilder.newBuilder().sslContext(sslcontext) not preserved with restfulWS-3.0
**Describe the bug**  
We moved to restfulWS-3.0 feature and from what I understand there is a replacement of CXF with RESTEASY. We use to set our own sslContext using ClientBuilder.newBuilder().sslContext(sslcontext).hostnameVerifier(...);.  This was working well with CXF but is seems that moving to RESTEASY our sslContext is ignored and replaced by the default ssl config defined in server.xml (replacement done during build() method invocation.

No stack trace associated just an unexpected behavior.

**Steps to Reproduce**  
```
        SSLContext sslcontext = SSLContext.getInstance("TLSv1.2");
        sslcontext.init(null, new TrustManager[] { new X509TrustManager() {
          @Override
          public void checkClientTrusted(X509Certificate[] arg0, String arg1)
              throws CertificateException {
          }
          @Override
          public void checkServerTrusted(X509Certificate[] arg0, String arg1)
              throws CertificateException {
          }
          @Override
          public X509Certificate[] getAcceptedIssuers() {
            return new X509Certificate[0];
          }

        } }, null);
        ClientBuilder builder = ClientBuilder.newBuilder();
        builder.sslContext(sslcontext).hostnameVerifier((s1, s2) -> true);
        // sslcontext defined in builded is correct
        Client client = builder.build();
        // after build sslcontext defined in builded is reset to default present in server.xml
```

**Expected behavior**  
 We would like sslcontext specified in builder to be used so that our trustmanager is invoked instead of server.xml default one.

**Diagnostic information:**  
 - OpenLiberty Version: 22.0.0.9/wlp-1.0.68.cl220920220815-1900
 - Affected feature(s) restfulWS-3.0
 - Java Version: Eclipse OpenJ9 VM, version 11.0.15-ea+10

**Additional context**  
Wondering if pb is there [https://github.com/OpenLiberty/open-liberty/blob/a2783947b1bf9afb184cb930a8bc60fc3[…]/restfulWS/internal/ssl/component/SslClientBuilderListener.java](https://github.com/OpenLiberty/open-liberty/blob/a2783947b1bf9afb184cb930a8bc60fc3c444f43/dev/io.openliberty.restfulWS.internal.ssl/src/io/openliberty/restfulWS/internal/ssl/component/SslClientBuilderListener.java#L62)
if no sslRef is defined or not found then jsseHelper.getSSLContext(sslRef, null, null) will return default ssl context and will override ssl buidler context. One option could be to use jsseHelper.getSSLContext method with tryDefault boolean that allows to control this default behavior

### LINKED ISSUE #23189: JAX-RS Client requests do not honor custom SSL cipher suites
We can [configure SSL](https://openliberty.io/docs/latest/reference/config/ssl.html) to enable custom cipher suites, however, `jaxrs-2.0` and `jaxrs-2.1` do not honor them when sending a client request with SSL configured.

### PR DESCRIPTION
**Pull request by WhiteCat22**
Fixes: #23189
Fixes: #22738

### CODE REVIEW

**Comment by WhiteCat22** [comment]
#libby #build

**Review by acdemyers — DISMISSED** [review]

File: dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java [hunk 1]
```diff
@@ -127,26 +130,52 @@ private TLSClientParameters retriveHTTPTLSClientParametersUsingSSLRef(HTTPCondui
...
+    }
+
+    private Object[] getJaxRsSSLManagerParams(String sslRef, Message message) {
+        String uriString = (String) message.get(Message.REQUEST_URI);
+        URI uri = URI.create(uriString);
+        int port = uri.getPort();
+
+        // if the port wasn't specified, use the default SSL port (443)
+        if (port == -1 && !uriString.contains(":-1")) {
+            port = 443;
```
Comment (jim-krueger) [inline dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java:166]: How about putting a tr.debug statement here saying that the port wasn't specified.

File: dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java [hunk 2]
```diff
@@ -156,21 +185,55 @@ public Method run() throws NoSuchMethodException, SecurityException {
...
+    }
 
-            Object[] parameters = { sslRef, uri.getHost(), Integer.toString(port) };
-            SSLSocketFactory ssLSocketFactory = (SSLSocketFactory) m.invoke(classObject, parameters);
-            return ssLSocketFactory;
+    private List<String> getSSLProtocols(String sslRef, Message message) {
+        try {
+            final Class<?> jaxrsSslMgrClass = getJaxRsSSLManagerClass();
+            Object classObject = jaxrsSslMgrClass.newInstance();
+            Method m = AccessController.doPrivileged(new PrivilegedExceptionAction<Method>() {
```
Comment (jim-krueger) [inline dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java:221]: Question:  Is this doPriv needed for Java 2 security only?   If so it should be gated by an if check.

File: dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java [hunk 3]
```diff
@@ -110,4 +76,138 @@ public Boolean run() throws SSLException {
...
+        }
+
+        return sslProps;
+    }
+
+    private static SSLContext getSSLContext(String sslRef, Map<String, Object> connectionInfo, Properties sslProps) throws SSLException {
+        Boolean sslCfgExists = null;
+        if (sslRef != null) {
+            try {
+                sslCfgExists = AccessController.doPrivileged(new PrivilegedExceptionAction<Boolean>() {
```
Comment (jim-krueger) [inline dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java:203]: Same question about the doPriv

File: dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java [hunk 4]
```diff
@@ -110,4 +76,138 @@ public Boolean run() throws SSLException {
...
+        connectionInfo.put(Constants.CONNECTION_INFO_DIRECTION, Constants.DIRECTION_OUTBOUND);
+        connectionInfo.put(Constants.CONNECTION_INFO_REMOTE_HOST, host);
+        connectionInfo.put(Constants.CONNECTION_INFO_REMOTE_PORT, port); // String expected by OutboundSSLSelections
+        return connectionInfo;
+    }
+
+    private static Properties getSSLProperties(String sslRef, Map<String, Object> connectionInfo) throws SSLException {
+        Properties sslProps;
+        try {
+            sslProps = AccessController.doPrivileged(new PrivilegedExceptionAction<Properties>() {
```
Comment (jim-krueger) [inline dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java:184]: Same question about doPriv

File: dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java (hunk 2 above)
Comment (WhiteCat22) [reply dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java:221]: dunno, this is how we get the SSLSocketFactory already so I'm just mirroring that. I'm hesitant to make any changes.

**Comment by WhiteCat22** [comment]
#libby #build

**Review by jim-krueger — DISMISSED** [review]

**Comment by WhiteCat22** [comment]
#libby #build

**Review by jim-krueger — DISMISSED** [review]

**Comment by WhiteCat22** [comment]
Made a minor update to return `null` for the `SSLSocketFactory` if the `SSLContext` is `null` to avoid an NPE.

**Comment by WhiteCat22** [comment]
#libby #build

File: dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java [hunk 5]
```diff
@@ -81,28 +83,29 @@ public void handleMessage(Message message) throws Fault {
...
+            TLSClientParameters tlsClientParams = retriveHTTPTLSClientParametersUsingSSLRef(httpConduit, sslRef, disableCNCheck, socketFactory, message);
             if (null != tlsClientParams) {
                 httpConduit.setTlsClientParameters(tlsClientParams);
             }
         }
     }
 
-    private TLSClientParameters retriveHTTPTLSClientParametersUsingSSLRef(HTTPConduit httpConduit, String sslRef, boolean disableCNCheck, SSLSocketFactory sslSocketFactory) {
+    private TLSClientParameters retriveHTTPTLSClientParametersUsingSSLRef(HTTPConduit httpConduit, String sslRef, boolean disableCNCheck, SSLSocketFactory sslSocketFactory,
+                                                                          Message message) {
```
Comment (jim-krueger) [inline dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java:108]: Eliminate this extra space?

**Review by jim-krueger — COMMENTED** [review]
One very small change

File: dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java (hunk 5 above)
Comment (jim-krueger) [reply dev/com.ibm.ws.jaxrs.2.0.client/src/com/ibm/ws/jaxrs20/client/security/LibertyJaxRsClientSSLOutInterceptor.java:108]: Spacing was off on review panel, but looks fine here.  So ignore.

**Review by jim-krueger — APPROVED** [review]

**Pull request merged**